File: app/src/main/cpp/wn-steam-client/src/cm_server_list.cpp
#include "wn_steam/cm_server_list.h"

#include <algorithm>

namespace wn_steam {

CmServerList::CmServerList() : bad_memory_(kDefaultBadMemory) {}

CmServerList::CmServerList(std::chrono::minutes bad_memory)
    : bad_memory_(bad_memory) {}

void CmServerList::replace_all(std::span<const CmServer> servers) {
    std::lock_guard<std::mutex> lk(mu_);
    entries_.clear();
    entries_.reserve(servers.size());
    for (const auto& s : servers) {
        entries_.push_back(Entry{s, ServerQuality::Good, {}});
    }
    next_index_ = 0;
}

void CmServerList::add(const CmServer& server) {
    std::lock_guard<std::mutex> lk(mu_);
    entries_.push_back(Entry{server, ServerQuality::Good, {}});
}

size_t CmServerList::size() const {
    std::lock_guard<std::mutex> lk(mu_);
    return entries_.size();
}

void CmServerList::promote_expired_locked() {
    const auto now = std::chrono::steady_clock::now();
    for (auto& e : entries_) {
        if (e.quality == ServerQuality::Bad && now - e.marked_bad_at >= bad_memory_) {
            e.quality = ServerQuality::Good;
            e.marked_bad_at = {};
        }
    }
}

std::optional<CmServer> CmServerList::next_good() {
    std::lock_guard<std::mutex> lk(mu_);
    if (entries_.empty()) return std::nullopt;

    promote_expired_locked();

    const size_t n = entries_.size();
    for (size_t i = 0; i < n; ++i) {
        const size_t idx = (next_index_ + i) % n;
        if (entries_[idx].quality == ServerQuality::Good) {
            next_index_ = (idx + 1) % n;
            return entries_[idx].server;
        }
    }
    return std::nullopt;  // every server is currently Bad
}

void CmServerList::mark_bad(std::string_view endpoint) {
    std::lock_guard<std::mutex> lk(mu_);
    for (auto& e : entries_) {
        if (e.server.endpoint == endpoint) {
            e.quality = ServerQuality::Bad;
            e.marked_bad_at = std::chrono::steady_clock::now();
            return;
        }
    }
}

void CmServerList::mark_good(std::string_view endpoint) {
    std::lock_guard<std::mutex> lk(mu_);
    for (auto& e : entries_) {
        if (e.server.endpoint == endpoint) {
            e.quality = ServerQuality::Good;
            e.marked_bad_at = {};
            return;
        }
    }
}

void CmServerList::reset_quality() {
    std::lock_guard<std::mutex> lk(mu_);
    for (auto& e : entries_) {
        e.quality = ServerQuality::Good;
        e.marked_bad_at = {};
    }
}

// ---------------------------------------------------------------------------
// SteamKit-style hardcoded fallback list. Used only when Steam Directory is
// unreachable. These hosts are advertised by Steam as long-lived WebSocket
// CMs; pick a handful across geographies so a first-launch user without
// network for the directory call still has something to try.
//
// Source of names: cross-referenced against ValvePython/steam's
// `steam.client.cm.bootstrap_cm_list` and JavaSteam's `SmartCMServerList`
// hard defaults (cmp1-sea1.steamserver.net:443 etc.). Steam rotates these
// over years; treat the list as best-effort.
// ---------------------------------------------------------------------------

std::vector<CmServer> hardcoded_fallback_servers() {
    // Steam advertises CM endpoints under the steamserver.net zone, in the
    // pattern `ext{N}-{dc}.steamserver.net:443` for WSS. Verified via real
    // GetCMListForConnect responses (see SteamKit2 + node-steam-user logs).
    // The previous *.cm.steampowered.com names did NOT resolve on tested
    // devices. These do.
    static constexpr struct {
        const char* host;
        const char* dc;
    } kFallback[] = {
        {"ext1-sea1.steamserver.net", "sea1"},
        {"ext2-sea1.steamserver.net", "sea1"},
        {"ext1-iad1.steamserver.net", "iad1"},
        {"ext2-iad1.steamserver.net", "iad1"},
        {"ext1-fra1.steamserver.net", "fra1"},
        {"ext2-fra1.steamserver.net", "fra1"},
        {"ext1-lax1.steamserver.net", "lax1"},
        {"ext1-sgp1.steamserver.net", "sgp1"},
    };

    std::vector<CmServer> out;
    out.reserve(sizeof(kFallback) / sizeof(kFallback[0]));
    for (const auto& f : kFallback) {
        CmServer s;
        s.host        = f.host;
        s.port        = 443;
        s.endpoint    = std::string(f.host) + ":443";
        s.transport   = CmTransport::WebSocket;
        s.realm       = "steamglobal";
        s.datacenter  = f.dc;
        s.load        = 0;       // unknown
        s.weighted_load = 0.0f;
        out.push_back(std::move(s));
    }
    return out;
}

}  // namespace wn_steam

File: app/src/main/cpp/wn-steam-client/rust/src/cm_server_list.rs
use crate::cm_server::{CmServer, CmTransport};
use std::sync::Mutex;
use std::time::{Duration, Instant};

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ServerQuality {
    Good,
    Bad,
}

pub const DEFAULT_BAD_MEMORY: Duration = Duration::from_secs(5 * 60);

pub struct CmServerList {
    inner: Mutex<Inner>,
    bad_memory: Duration,
}

#[derive(Default)]
struct Inner {
    entries: Vec<Entry>,
    next_index: usize,
}

struct Entry {
    server: CmServer,
    quality: ServerQuality,
    marked_bad_at: Option<Instant>,
}

impl Default for CmServerList {
    fn default() -> Self {
        Self::new(DEFAULT_BAD_MEMORY)
    }
}

impl CmServerList {
    pub fn new(bad_memory: Duration) -> Self {
        Self {
            inner: Mutex::new(Inner::default()),
            bad_memory,
        }
    }

    pub fn replace_all(&self, servers: &[CmServer]) {
        let mut inner = self.inner.lock().unwrap();
        inner.entries = servers
            .iter()
            .cloned()
            .map(|server| Entry {
                server,
                quality: ServerQuality::Good,
                marked_bad_at: None,
            })
            .collect();
        inner.next_index = 0;
    }

    pub fn add(&self, server: CmServer) {
        self.inner.lock().unwrap().entries.push(Entry {
            server,
            quality: ServerQuality::Good,
            marked_bad_at: None,
        });
    }

    pub fn size(&self) -> usize {
        self.inner.lock().unwrap().entries.len()
    }

    pub fn next_good(&self) -> Option<CmServer> {
        let mut inner = self.inner.lock().unwrap();
        if inner.entries.is_empty() {
            return None;
        }
        promote_expired(&mut inner.entries, self.bad_memory);
        let n = inner.entries.len();
        for i in 0..n {
            let idx = (inner.next_index + i) % n;
            if inner.entries[idx].quality == ServerQuality::Good {
                inner.next_index = (idx + 1) % n;
                return Some(inner.entries[idx].server.clone());
            }
        }
        None
    }

    pub fn mark_bad(&self, endpoint: &str) {
        let mut inner = self.inner.lock().unwrap();
        if let Some(entry) = inner
            .entries
            .iter_mut()
            .find(|e| e.server.endpoint == endpoint)
        {
            entry.quality = ServerQuality::Bad;
            entry.marked_bad_at = Some(Instant::now());
        }
    }

    pub fn mark_good(&self, endpoint: &str) {
        let mut inner = self.inner.lock().unwrap();
        if let Some(entry) = inner
            .entries
            .iter_mut()
            .find(|e| e.server.endpoint == endpoint)
        {
            entry.quality = ServerQuality::Good;
            entry.marked_bad_at = None;
        }
    }

    pub fn reset_quality(&self) {
        let mut inner = self.inner.lock().unwrap();
        for entry in &mut inner.entries {
            entry.quality = ServerQuality::Good;
            entry.marked_bad_at = None;
        }
    }
}

fn promote_expired(entries: &mut [Entry], bad_memory: Duration) {
    let now = Instant::now();
    for entry in entries {
        if entry.quality == ServerQuality::Bad
            && entry
                .marked_bad_at
                .is_some_and(|t| now.duration_since(t) >= bad_memory)
        {
            entry.quality = ServerQuality::Good;
            entry.marked_bad_at = None;
        }
    }
}

pub fn hardcoded_fallback_servers() -> Vec<CmServer> {
    const FALLBACK: &[(&str, &str)] = &[
        ("ext1-sea1.steamserver.net", "sea1"),
        ("ext2-sea1.steamserver.net", "sea1"),
        ("ext1-iad1.steamserver.net", "iad1"),
        ("ext2-iad1.steamserver.net", "iad1"),
        ("ext1-fra1.steamserver.net", "fra1"),
        ("ext2-fra1.steamserver.net", "fra1"),
        ("ext1-lax1.steamserver.net", "lax1"),
        ("ext1-sgp1.steamserver.net", "sgp1"),
    ];
    FALLBACK
        .iter()
        .map(|(host, dc)| CmServer {
            endpoint: format!("{host}:443"),
            host: (*host).to_string(),
            port: 443,
            transport: CmTransport::WebSocket,
            realm: "steamglobal".to_string(),
            datacenter: (*dc).to_string(),
            load: 0,
            weighted_load: 0.0,
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn next_good_round_robins_and_skips_bad() {
        let list = CmServerList::default();
        let a = server("a:443");
        let b = server("b:443");
        list.replace_all(&[a.clone(), b.clone()]);
        assert_eq!(list.next_good().unwrap().endpoint, a.endpoint);
        list.mark_bad(&b.endpoint);
        assert_eq!(list.next_good().unwrap().endpoint, a.endpoint);
        list.mark_good(&b.endpoint);
        assert_eq!(list.next_good().unwrap().endpoint, b.endpoint);
    }

    #[test]
    fn bad_servers_promote_after_memory_interval() {
        let list = CmServerList::new(Duration::ZERO);
        let a = server("a:443");
        list.replace_all(std::slice::from_ref(&a));
        list.mark_bad(&a.endpoint);
        assert_eq!(list.next_good().unwrap().endpoint, a.endpoint);
    }

    #[test]
    fn fallback_servers_are_websocket_urls() {
        let servers = hardcoded_fallback_servers();
        assert!(!servers.is_empty());
        assert!(servers
            .iter()
            .all(|s| s.websocket_url().starts_with("wss://")));
    }

    fn server(endpoint: &str) -> CmServer {
        let (host, port) = crate::cm_server::parse_endpoint(endpoint).unwrap();
        CmServer {
            endpoint: endpoint.to_string(),
            host,
            port,
            transport: CmTransport::WebSocket,
            ..Default::default()
        }
    }
}
